File: qqbot-plugin-api/src/main/java/com/mieai/qqbot/plugin/api/MessageSender.java
package com.mieai.qqbot.plugin.api;

import java.util.concurrent.CompletionStage;

/** Controlled capability; implementations enqueue durable Outbox work, never expose tokens. */
@FunctionalInterface
public interface MessageSender {
    CompletionStage<MessageEnqueueReceipt> enqueue(TextMessage message);

    default CompletionStage<MessageEnqueueReceipt> enqueue(MediaMessage message) {
        return java.util.concurrent.CompletableFuture.failedFuture(
                new UnsupportedOperationException("media messages are not supported by this sender"));
    }

    default CompletionStage<MessageEnqueueReceipt> enqueue(RichMessage message) {
        return java.util.concurrent.CompletableFuture.failedFuture(
                new UnsupportedOperationException("rich messages are not supported by this sender"));
    }
}

File: qqbot-plugin-api/src/main/kotlin/com/mieai/qqbot/plugin/api/MessageSender.kt
package com.mieai.qqbot.plugin.api

import java.util.UUID
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage

/** Controlled capability; implementations enqueue durable Outbox work, never expose tokens. */
interface MessageSender {
    fun enqueue(message: TextMessage): CompletionStage<MessageEnqueueReceipt>

    fun enqueue(
        message: TextMessage,
        options: MessageSendOptions,
    ): CompletionStage<MessageEnqueueReceipt> = withDefaultOptions(options) { enqueue(message) }

    fun enqueue(message: MediaMessage): CompletionStage<MessageEnqueueReceipt>

    fun enqueue(
        message: MediaMessage,
        options: MessageSendOptions,
    ): CompletionStage<MessageEnqueueReceipt> = withDefaultOptions(options) { enqueue(message) }

    fun enqueue(message: RichMessage): CompletionStage<MessageEnqueueReceipt>

    fun enqueue(
        message: RichMessage,
        options: MessageSendOptions,
    ): CompletionStage<MessageEnqueueReceipt> = withDefaultOptions(options) { enqueue(message) }

    /**
     * Reads the durable lifecycle and QQ delivery result for a previously enqueued message.
     * Results are scoped to the current plugin binding and remain queryable after plugin restart.
     */
    fun findDelivery(jobId: UUID): CompletionStage<MessageDeliveryReceipt?>
}

private fun <T> withDefaultOptions(options: MessageSendOptions, enqueue: () -> CompletionStage<T>): CompletionStage<T> =
    if (options.messageReference == null) {
        enqueue()
    } else {
        CompletableFuture.failedFuture(
            UnsupportedOperationException("This message sender does not support explicit message references"),
        )
    }
